Proposal to add tykky shell convenience function to activate and deactivate environments

xavierabellan opened the pull request:
This is a proposal to add a `tykky` shell function so that wrapped environments can be easily "activated" and "deactivated", borrowing the concept from conda. Instead of manually adding the `<env_dir>/bin` to the `PATH`, those installations can easily be made available with `tykky activate <env>`. They can also be removed more cleanly from the environment with `tykky deactivate`.

As a bonus, and for added convenience:

- environments can also be activated by "name" if they are found in the colon-separated, `PATH`-like variable `TYKKY_PATH`. 
- If running interactively, mimicking what is done in conda or Python venvs, PS1 is updated prepending the environment name.

This feature makes adoption of the tool for existing conda users much smoother and natural, since they just need to replace `conda activate` by `tykky activate` in their workflows.

Note that this is an initial, very simple implementation of the concept, and only POSIX-compliant shells are supported (no csh). If it gains any traction, it may be worth reviewing in order to see how to further and better integrate with the rest of the tool i.e. defining defaults for TYKKY_PATH in the YAML config).

Nortamo commented:
Idea seems nice, and the basic feature works well, some refinement probably require for "before" and "after"

The base `tykky` command is now provided by a shell function, which are not by default exported to subshell, this might 
have implications on certain workflows and might be unintuitive for some users, e.g if conda activate works on the command line, it will work in an `sbatch` invocation but the `tykky` command would not be found. 

So as we are now just targeting bash one could export the function to have it be available in subshells ?

The printed message is now a bit confusing as users who just loaded a module to do an installation might not have the `tykky` function defined (as it would at this point require manually sourcing the file), perhaps change it so that the hint is only shown if the command is available ? 

What's the vision for using this? 
1. Providing instructions for users to manually source the  etc/profile.d/tykky file when they need to activate an environment ? 
2. Having the same module (or a separate module ) provide the shell function for activating the environment -> so users load a module when they need to do installations or use existing ones ? (Note that in the case of lmod functions defined cannot be exported -> wont work in non-interactive subshells, `#!/bin/bash -l` all your scripts and sbatch scripts  or load the module again explicitly. In this case it might sense to have tykky be a script on disk instead )
3. Going the conda way and doing one time magic in `.bashrc` (personally I very much dislike touching user's bashrc)
4. placing a tykky binary in e.g `~/.local/bin` -> interesting issues if running multiple versions of Tykky if the logic changes drastically (unlikely)

There are other features / questions on how or if to populate TYKKY path or again just instructing the users. 

I don't have a strong preference towards any option.  

But I think the basic idea is nice and feels familiar to users, for this PR specifically I think it would be sufficient, to hide the prompt if `tykky` command is not available (or make it a toggle option in the config), and possibly export the function / have it as a binary. 

The it can be left for sites to decide exactly how to expose the activation functionality (any of the listed options) to end users and/or  we can implement some more quality of life features in a future PR. 

Or then if you feel strongly for some specific option let me know, and I'll think if that makes sense as default behavior ?

Nortamo commented:
Pushed a fix, so now the function is exported and I hide the prompt if the command is not available. 
Would this work for the envisioned use case.

xavierabellan commented:
Thank you! it also looks good from this side, so you may go ahead and merge

xavierabellan commented:
After a few tests, I realised that ksh support was broken. I refactored the shell function definition so that both bash and ksh are now working, including the cases when called from subshells or scripts.

For completeness, and as mentioned in a previous comment, I also added the PATH definition when missing so that users whishing to use the tool on unmanaged platforms can just source the etc/profile.d/tykky.sh for the full experience.

Nortamo commented:
LGTM, I'll do some final checks, and if OK will do a merge. After this and #25  have been merged I'll tag a new 0.4-something release.

xavierabellan commented:
Thanks!

Nortamo commented:
I added an error message to give a hint to the user, i.e if the folder exists but is not valid vs if we could not find it at all. 

Also for consistency I allowed the activation of environments in the current directory as we now allow activation via relative paths so it could be confusing if e.g `tykky activate folder/my_env` works but `cd folder && tykky activate my_env` does not. Also added a warning to cover the edge case where we find multiple candidates both from `TYKKY_PATH` and from the current directory. 

It's a bit verbose, but should help support when users get lazy and reuse names for environments in different places.

If you think these final additions are  ok I could merge this.

xavierabellan commented:
Looks good to me too, it does make it a bit more robust. I just pushed an additional commit removing some echo's which were left by mistake in tykky.sh
